Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp

```diff
@@ -0,0 +1,366 @@
+
+  Workspace_sptr inputWS = getProperty("InputWorkspace");
+
+  checkDetectorIDsInWorkspace(options.grouping, inputWS);
+
+  WorkspaceGroup_sptr groupedWS = getProperty("WorkspaceGroup");
+  if (!groupedWS) {
+    groupedWS = addGroupedWSWithDefaultName(inputWS);
+  }
+
+  // Perform the analysis specified in options
+  addGroupingToADS(options, inputWS, groupedWS);
+  addPairingToADS(options, inputWS, groupedWS);
+
+  if (getProperty("ApplyAsymmetryToGroups")) {
+    options.plotType = PlotType::Asymmetry;
+    addGroupingToADS(options, inputWS, groupedWS);
+  }
+  addGroupingInformationToADS(options.grouping);
+}
+
+// Checks that the detector IDs in grouping are in the workspace
+void LoadAndApplyMuonDetectorGrouping::checkDetectorIDsInWorkspace(
+    API::Grouping &grouping, Workspace_sptr workspace) {
+  bool check =
+      MuonAlgorithmHelper::checkGroupDetectorsInWorkspace(grouping, workspace);
+  if (!check) {
+    g_log.error("One or more detector IDs specified in the groups is not "
+                "contained in the InputWorkspace");
+    throw std::runtime_error(
+        "One or more detector IDs specified in the groups is not "
+        "contained in the InputWorkspace");
+  }
+}
+
+/**
+ * Adds an empty WorkspaceGroup to the ADS with a name
+ * that it would have if created by the MuonAnalysis GUI.
+ */
+WorkspaceGroup_sptr
+LoadAndApplyMuonDetectorGrouping::addGroupedWSWithDefaultName(
+    Workspace_sptr workspace) {
+  auto &ads = AnalysisDataService::Instance();
+  std::string groupedWSName = MuonAlgorithmHelper::getRunLabel(workspace);
+
+  WorkspaceGroup_sptr groupedWS;
+  if (ads.doesExist(groupedWSName)) {
+    if ((groupedWS = boost::dynamic_pointer_cast<WorkspaceGroup>(
+             ads.retrieve(groupedWSName)))) {
+      return groupedWS;
+    }
+  }
+
+  groupedWS = boost::make_shared<WorkspaceGroup>();
+  AnalysisDataService::Instance().addOrReplace(groupedWSName, groupedWS);
+  return groupedWS;
+}
+
+// Set the parameters in options to sensible defaults for this algorithm
+AnalysisOptions LoadAndApplyMuonDetectorGrouping::setDefaultOptions() {
+  AnalysisOptions options;
+  options.summedPeriods = this->getPropertyValue("SummedPeriods");
+  options.subtractedPeriods = this->getPropertyValue("SubtractedPeriods");
+  options.timeZero = 0.0;
+  options.loadedTimeZero = this->getProperty("TimeOffset");
+  options.rebinArgs = this->getPropertyValue("RebinArgs");
+  options.plotType = Muon::PlotType::Counts;
+  return options;
+}
+
+/**
+ * Adds the group names and corresponding detector IDs to
+ * a TableWorkspace in the ADS named "MuonGroupings"
+ */
+void LoadAndApplyMuonDetectorGrouping::addGroupingInformationToADS(
+    const Mantid::API::Grouping &grouping) {
+
+  auto groupingTable = boost::dynamic_pointer_cast<ITableWorkspace>(
+      WorkspaceFactory::Instance().createTable("TableWorkspace"));
+  groupingTable->addColumn("str", "GroupName");
+  groupingTable->addColumn("vector_int", "Detectors");
```

Comment:
you probably also want to store the run number, so you know which group corresponds to which run

Reply:
I think it's best to hold off on these modifications for now. I've added the option to produce the table and defaulted it to false.

If this table needs to be used during scripting then there is a substantial amount of functionality that would be added which is probably best for another PR; e.g. having to update the table, store multiple tables for multiple runs, updating the other muon grouping/pairing algorithms.

Reply:
can we have an issue for this?